# Part 1
## Step 2
## Importing libraries and customer data set as dataframe


In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## Step 3

In [2]:
# Creating path
path = r'C:\Users\terry\Documents\Data Analytics Course\Achievement 4\23-11-2024 Instacart Basket Analysis'

In [3]:
# importing customers dataset as dataframe
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

## Step 4
## Customer Dataframe Wrangling 

In [4]:
# Checking customers dataframe for possible renaming and dropping of unnecessary columns
df.head(1)

user_id First Name    Surnam  Gender     STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665

In [5]:
# Checking data dimensions
df.shape

(206209, 10)

#### Renaming illogical column names


In [6]:
df = df.rename(columns = {
    'First Name' : 'first_name',
    'Surnam' : 'last_name',
    'Gender' : 'gender',
    'STATE': 'state',
    'Age' : 'age',
    'n_dependants' : 'number_of_dependents',
    'fam_status': 'family_status'})

In [7]:
# Checking renamed columns
df.head(1)

user_id first_name last_name  gender     state  age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   number_of_dependents family_status  income  
0                     3       married  165665

## Step 5
## Data quality and Consistency checks

In [8]:
# Checking Statistical analysis of customers dataframe
df.describe()

user_id            age  number_of_dependents         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

Despite the max income value of 593901, nothing seems out of the ordinary and couldn't be considered an outlier

In [9]:
# Searching missing values in customers dataframe
missing_values = df.isnull().sum()

In [10]:
# Checking missing values
missing_values

user_id                     0
first_name              11259
last_name                   0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependents        0
family_status               0
income                      0
dtype: int64

Despite a huge number of missing first names, this doesn't warrant deleting any data. I do however question the missing first names.

In [11]:
# Searching for duplicate observations
duplicates = df.duplicated()

In [12]:
# Total duplicates
num_duplicates = duplicates.sum()

In [13]:
num_duplicates

0

There are no duplicates in the customers dataframe

In [14]:
# Checking for mixed data types
df.dtypes

user_id                  int64
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependents     int64
family_status           object
income                   int64
dtype: object

Data types are consistent with column variables

## Step 6
### Combining customers data with prepared Instacart data

In [15]:
# Importing Instacart prepared data
df_ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'aggregated_ords_prods_merged.pkl'))

In [19]:
# Checking df_ords_prods for common column
df_ords_prods.head(1)

order_id  user_id  number_of_order  orders_day_of_week  order_hour_of_day  \
0   2539329        1                1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   

  product_name  ...  prices  match    busiest_days busiest_period_of_day  \
0         Soda  ...     9.0   both  Regularly busy        Average orders   

  max_order  loyalty_flag  average_spend spender_flag  Customer_frequency  \
0        10  New customer       6.367797  Low spender                20.5   

          frequency_flag  
0  Non frequent customer  

[1 rows x 22 columns]

A merge is possible due to the 'user_id' common column on both dataframes

In [28]:
# Checking if common column 'user_id' have the same data types
df_ords_prods.dtypes

order_id                     int64
user_id                      int64
number_of_order              int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
match                     category
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
average_spend              float64
spender_flag                object
Customer_frequency         float64
frequency_flag              object
dtype: object

'user_id' on both dataframes have the same data type (integer) and it's not necessary to change either or both to a string.

In [16]:
# Merging df_ords_prods with customers dataframe on 'user_id'
ords_prods_cust_merged = pd.merge(df_ords_prods, df, on = 'user_id', how = 'inner')

In [22]:
pd.options.display.max_columns = None

In [24]:
# Checking merged Dataframes
ords_prods_cust_merged.head()

order_id  user_id  number_of_order  orders_day_of_week  order_hour_of_day  \
0   2539329        1                1                   2                  8   
1   2539329        1                1                   2                  8   
2   2539329        1                1                   2                  8   
3   2539329        1                1                   2                  8   
4   2539329        1                1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  match    busiest_days busiest_period_of_day  max_order  loyalty_flag  \
0  both  Regularly busy        Average orders         10  New customer   
1  both  Regularly busy        Average orders         10  New customer   
2  both  Regularly busy        Average orders         10  New customer   
3  both  Regularly busy        Average orders         10  New customer   
4  both  Regularly busy        Average orders         10  New customer   

   average_spend spender_flag  Customer_frequency         frequency_flag  \
0       6.367797  Low spender                20.5  Non frequent customer   
1       6.367797  Low spender                20.5  Non frequent customer   
2       6.367797  Low spender                20.5  Non frequent customer   
3       6.367797  Low spender                20.5  Non frequent customer   
4       6.367797  Low spender                20.5  Non frequent customer   

  first_name last_name  gender    state  age date_joined  \
0      Linda    Nguyen  Female  Alabama   31   2/17/2019   
1      Linda    Nguyen  Female  Alabama   31   2/17/2019   
2      Linda    Nguyen  Female  Alabama   31   2/17/2019   
3      Linda    Nguyen  Female  Alabama   31   2/17/2019   
4      Linda    Nguyen  Female  Alabama   31   2/17/2019   

   number_of_dependents family_status  income  
0                     3       married   40423  
1                     3       married   40423  
2                     3       married   40423  
3                     3       married   40423  
4                     3       married   40423

In [20]:
ords_prods_cust_merged.shape

(32404859, 31)

## Step 8
### Exporting merged dataframe

In [26]:
ords_prods_cust_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customers_merged.pkl'))